Implement SquareSegmentedProgressIndicator

oas004 opened the pull request:
#### WHAT
Here I have tried to implement the square version of the `SegmentedProgressIndicator` . 

#### WHY
Because of #431 

#### HOW
I have made the API similar to `SegmentedProgressIndicator`. The difficulties on this comes from painting the rounded square. There is no method in the Canvas to draw a progress of a rounded square shape. I have tried to do this by combining `drawArch` and `drawLine` based on these sections.

Now, the implementation is too tied to having a sum of 8 on the section weight. I need to decouple that somehow. I also kinda want the stroke cap to be round. If I can't have one end round and one square, I am unsure if I can make that possible with this approach. The reason is that then you would be able to see the original 8 sections even though they are supposed to be combined into one section.


#### Checklist :clipboard:
- [x] Add explicit visibility modifier and explicit return types for public declarations
- [x] Run spotless check
- [x] Run tests
- [x] Update metalava's signature text files


![image](https://user-images.githubusercontent.com/231923/201885586-b07a8ff2-bc76-4e6f-ae9b-2241445f6d6e.png)

yschimke commented:
Sorry, I was on vacation last week. I'll take a look in next couple of days.

yschimke commented:
It looks really good.  I'm still working through it and might put up some changes for your PR.

Main feedback

- The anim preview is great, but should use updateTransition so preview tooling works

```
    val transition = updateTransition(
        targetState = PreviewAnimationState.Start,
        label = "Square Progress Indicator"
    )

    val progress by transition.animateFloat(
        label = "Progress",
        targetValueByState = {
            it.target
        },
        transitionSpec = {
            tween(durationMillis = 1000, easing = LinearEasing)
        }
    )
``` 

- We should follow the conventions from SegmentedProgressIndicator, so maybe use ProgressIndicatorSegment as the segment types, add trackColor as a param.

- Define progressSemantics on Canvas

- The padding 38 hides drawing outside the canvas I think.

yschimke commented:
I'm going to try a different strategy.  Basically artificially inserting the gaps as artificial segments, drawing as the 8 adjacent lines/arcs. And for each line/arc drawing a subset based on the progress through my computed segments.

oas004 commented:
> It looks really good. I'm still working through it and might put up some changes for your PR.
> 
> Main feedback
> 
> * The anim preview is great, but should use updateTransition so preview tooling works
> 
> ```
>     val transition = updateTransition(
>         targetState = PreviewAnimationState.Start,
>         label = "Square Progress Indicator"
>     )
> 
>     val progress by transition.animateFloat(
>         label = "Progress",
>         targetValueByState = {
>             it.target
>         },
>         transitionSpec = {
>             tween(durationMillis = 1000, easing = LinearEasing)
>         }
>     )
> ```
> 
>  * We should follow the conventions from SegmentedProgressIndicator, so maybe use ProgressIndicatorSegment as the segment types, add trackColor as a param.
> * Define progressSemantics on Canvas
> * The padding 38 hides drawing outside the canvas I think.
> * for comparing against SegmentedProgressIndicator, we should reuse same colours.

Sorry for late reply, my girlfriend had a birthday yesterday I had to attend. Thanks for awesome suggestions!  I have in 
0cdb56d moved the previews out to the debug source set, added the updateTransition and made sure that they work in the preview tooling as well :) I have also changed the colors and the api to match the `SegmentedProgressIndicator`. I also added the progressSemantics modifier to the Canvas composable (is that correct?) :) For the padding, I think you are correct. Is that something that should be addressed in the drawing algorithm? Or should the user use the modifier to have whatever the padding they want? :)

The strategy sounded kinda advanced, but I think it is a great idea if it works! Feel free to hijack the PR if that is easier :)

yschimke commented:
No plan to hijack, wanted to present some options.

oas004 commented:
> No plan to hijack, wanted to present some options.

Alright :) I had some sleep on the idea and I was wondering; Is it possible to have like an inverse stroke cap? Or should the cap just be a rectangle like it is now? :)

yschimke commented:
Not sure. It ends up overlapping depending on how you configure the Stroke and padding.  But something to play with.

yschimke commented:
@oas004 let me know when it's ready to take a look.

oas004 commented:
> @oas004 let me know when it's ready to take a look.

Thank you man! I think I introduced one defect when there is a lot of segments. Will try to fix that and let you know😊

oas004 commented on composables/src/main/java/com/google/android/horologist/composables/SquareSegmentedProgressIndicator.kt:
```diff
@@ -0,0 +1,457 @@
+    trackColor: Color = MaterialTheme.colors.onBackground.copy(alpha = 0.1f),
+    cornerRadiusDp: Dp = 10.dp,
+    trackSegments: List<ProgressIndicatorSegment>,
+    paddingDp: Dp = ProgressIndicatorDefaults.StrokeWidth
```
Here the user can give a padding that is going to be in-between the segments (so the gap size), would it make sense to maybe make this an enum and have fixed sizes? Or say that it can't be bigger than 10 or something? If it is bigger than 10 it starts to almost not make sense. Just thinking out loud here, what do you guys think? :)  Maybe following the same as api as `SegmentedProgressIndicator` makes sense 🤔

yschimke replied:
+1, can't answer on this specific. But copying api and features from SPI makes sense.

oas004 replied:
I think the API from SPI is that the user can choose the padding angle which in that case is the equivalent of our spacing here. If they choose 360f then the entire circle is just a padding I would guess, and I think that might be a bit strange 🤔

oas004 commented:
@yschimke I think this is ready to be looked at. I took a lot of things from your branch https://github.com/yschimke/horologist/tree/sspb that was way better! That helped a lot! Thank you so much for that! 🙏

oas004 commented on composables/src/debug/java/com/google/android/horologist/composables/SquareSegmentedProgressIndicatorPreview.kt:
```diff
@@ -0,0 +1,196 @@
+@OptIn(ExperimentalHorologistComposablesApi::class)
+@WearSquareDevicePreview
+@Composable
+fun PreviewProgressAnimation() {
```
I think this got very nice now! 😄

yschimke replied:
Yep - it's awesome

oas004 commented on composables/src/main/java/com/google/android/horologist/composables/SquareSegmentedProgressIndicator.kt:
```diff
@@ -0,0 +1,458 @@
+    }
+
+    check(cornerRadiusDp in 0.dp..90.dp) {
+        "If your corner radius is above 90, you should use SegmentedProgressIndicator."
```
Is this enough? Or should I expand on it? :)

yschimke replied:
I'm not sure this is true. Perhaps we should constrain it based on strokeWidth? or make it a boolean for none/strokeWidth. It's not in degrees, so not sure 90 make sense here.

oas004 replied:
Yeah I think you are correct when I think about it. I guess this will depend on the size as well, so I will remove this check. I'm not sure I get 100% what you mean about constraining the corner radius based on StrokeWidth. Do you mean the API should be like this?

```
@ExperimentalHorologistComposablesApi
@Composable
public fun SquareSegmentedProgressIndicator(
    modifier: Modifier = Modifier,
    progress: Float,
    strokeWidth: Dp = ProgressIndicatorDefaults.StrokeWidth,
    trackColor: Color = MaterialTheme.colors.onBackground.copy(alpha = 0.1f),
    shouldHaveStrokeWidthCornerRadius: Boolean = false,
    trackSegments: List<ProgressIndicatorSegment>,
    paddingDp: Dp = ProgressIndicatorDefaults.StrokeWidth
)
``` 
and then setting it to none if it is false making it a square? :)

yschimke replied:
Sorry was mixing this up with the gap.  But the rest of my comment stands.

oas004 commented on composables/src/test/kotlin/com/google/android/horologist/composables/SquareSegmentedProgressIndicatorTest.kt:
```diff
@@ -0,0 +1,185 @@
+import org.junit.Rule
+import org.junit.Test
+
+class SquareSegmentedProgressIndicatorTest {
```
Do I need to add more Paparazzi tests? Or is this enough?

yschimke replied:
I think enough, expand more as we find a reason.

yschimke commented on composables/src/main/java/com/google/android/horologist/composables/SquareSegmentedProgressIndicator.kt:
```diff
@@ -0,0 +1,458 @@
+ * SquareSegmentedProgressIndicator represents a segmented progress indicator with
+ * square shape.
+ *
+ * @param modifier
```
nit: 
```suggestion
 * @param modifier The modifier to be applied to the layout.
```

oas004 replied:
Nice suggestion!Thanks for pointing it out! I added it to the [5985e44](https://github.com/google/horologist/pull/695/commits/5985e44b15ab7ce73f320a35006006496d35dabe) commit and rebased to update the branch. Hope that was okay :)

yschimke commented on composables/src/main/java/com/google/android/horologist/composables/SquareSegmentedProgressIndicator.kt:
```diff
@@ -0,0 +1,458 @@
+        "Only progress between 0.0 and 1.0 is allowed."
+    }
+
+    check(cornerRadiusDp in 0.dp..90.dp) {
```
I'm not sure this is true. Perhaps we should constrain it based on strokeWidth? or make it a boolean for none/strokeWidth.  It's not in degrees, so not sure 90 make sense here.
